general/mem_manager.hpp

```diff
@@ -394,158 +434,254 @@ class Memory
-   /// Check if pointer has been registered in the memory manager
-   void RegisterCheck(void *ptr);
+   /// Check if the host pointer has been registered in the memory manager
+   void RegisterCheck(void *h_ptr);
 
    /// Prints all pointers known by the memory manager
    void PrintPtrs(void);
+
+   /// Prints all aliases known by the memory manager
+   void PrintAliases(void);
+
+   MemoryType GetHostMemoryType() { return host_mem_type; }
+   MemoryType GetDeviceMemoryType() { return device_mem_type; }
 };
 
 
 // Inline methods
 
+template <typename T>
+inline void Memory<T>::Reset()
+{
+   h_ptr = NULL;
+   h_mt = MemoryManager::host_mem_type;
+   capacity = 0;
+   flags = 0;
+}
+
+template <typename T>
+inline void Memory<T>::Reset(MemoryType mt)
+{
+   h_ptr = NULL;
+   h_mt = mt;
+   capacity = 0;
+   flags = 0;
+}
+
+template <typename T>
+inline void Memory<T>::New(int size)
+{
+   capacity = size;
+   flags = OWNS_HOST | VALID_HOST;
+   h_mt = MemoryManager::host_mem_type;
+   h_ptr = (h_mt == MemoryType::HOST) ? new T[size] :
+           (T*)MemoryManager::New_(nullptr, size*sizeof(T), h_mt, flags);
+   MemoryManager::CheckHostMemoryType_(h_mt, h_ptr);
+}
+
 template <typename T>
 inline void Memory<T>::New(int size, MemoryType mt)
 {
-   if (mt == MemoryType::HOST)
-   {
-      New(size);
-   }
-   else
+   capacity = size;
+   if (mt == MemoryType::HOST) { flags = OWNS_HOST | VALID_HOST; }
+   h_mt = IsHostMemory(mt) ? mt : MemoryManager::GetDualMemoryType_(mt);
+   T *h_tmp = (h_mt == MemoryType::HOST) ? new T[size] : nullptr;
+   h_ptr = (mt == MemoryType::HOST) ? h_tmp:
+           (T*)MemoryManager::New_(h_tmp, size*sizeof(T), mt, flags);
+   MemoryManager::CheckHostMemoryType_(h_mt, h_ptr);
+}
+
+template <typename T>
+inline void Memory<T>::Wrap(T *host_ptr, int size, bool own)
+{
+   capacity = size;
+   h_ptr = host_ptr;
+   h_mt = MemoryType::HOST;
+   flags = (own ? OWNS_HOST : 0) | VALID_HOST;
+#ifdef MFEM_DEBUG
+   if (own && MemoryManager::Exists())
    {
-      // Allocate the host pointer with new T[] if 'mt' is a pure device memory
-      // type, e.g. CUDA.
-      T *tmp = (mt == MemoryType::CUDA) ? new T[size] : NULL;
-      h_ptr = (T*)MemoryManager::New_(tmp, size*sizeof(T), mt, flags);
-      capacity = size;
+      MFEM_VERIFY(!MemoryManager::IsKnown_(h_ptr), "Wrap but registered");
+      MFEM_VERIFY(!MemoryManager::IsAlias_(h_ptr), "Wrap but existing alias");
    }
+#endif
 }
 
 template <typename T>
 inline void Memory<T>::Wrap(T *ptr, int size, MemoryType mt, bool own)
 {
-   if (mt == MemoryType::HOST)
-   {
-      Wrap(ptr, size, own);
-   }
-   else
-   {
-      // Allocate the host pointer with new T[] if 'mt' is a pure device memory
-      // type, e.g. CUDA.
-      T *tmp = (mt == MemoryType::CUDA) ? new T[size] : NULL;
-      h_ptr = (T*)MemoryManager::Register_(ptr, tmp, size*sizeof(T), mt, own,
-                                           false, flags);
-      capacity = size;
-   }
+   h_mt = mt;
```

Comment:
I just noticed: here it is assumed `mt` is a host type but generally the idea is to support all host and device types.

Reply:
Right, I pulled the logic from `New` into `Wrap` too.